File: src/main/java/indi/fcwyzzr/minecraft/food_talks/common/data_component/compound_food/FoodItemProperties.java
package indi.fcwyzzr.minecraft.food_talks.common.data_component.compound_food;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.core.component.DataComponentType;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;
import net.neoforged.neoforge.registries.DeferredHolder;

import java.util.Optional;

import static indi.fcwyzzr.minecraft.food_talks.FoodTalksKt.toRegistryName;
import static indi.fcwyzzr.minecraft.food_talks.FoodTalksKt.toResourceLocation;

public record FoodItemProperties(
    boolean canAlwaysEat,
    Optional<ItemStack> convertsTo,
    float chewSeconds
) {
    public static final Codec<FoodItemProperties> codec = RecordCodecBuilder.create(
        it -> it.group(
            Codec.BOOL
                .fieldOf("can_always_eat")
                .forGetter(FoodItemProperties::canAlwaysEat),
            ItemStack.SINGLE_ITEM_CODEC
                .optionalFieldOf("using_converts_to")
                .forGetter(FoodItemProperties::convertsTo),
            Codec.FLOAT
                .fieldOf("eat_seconds")
                .forGetter(FoodItemProperties::chewSeconds)
        ).apply(it, FoodItemProperties::new)
    );
    public static final StreamCodec<RegistryFriendlyByteBuf, FoodItemProperties> streamCodec = StreamCodec
        .composite(
            ByteBufCodecs.BOOL,
            FoodItemProperties::canAlwaysEat,
            
            ItemStack.STREAM_CODEC.apply(ByteBufCodecs::optional),
            FoodItemProperties::convertsTo,
            
            ByteBufCodecs.FLOAT,
            FoodItemProperties::chewSeconds,
            
            FoodItemProperties::new
        );
    
    
    
    public static final String name = toRegistryName(FoodItemProperties.class.getSimpleName());
    
    public static final ResourceLocation location = toResourceLocation(name);
    
    public static DataComponentType<FoodItemProperties> type = DataComponentType
        .<FoodItemProperties>builder()
        .persistent(codec)
        .networkSynchronized(streamCodec)
        .build();
}

File: src/main/kotlin/indi/muxin/food_talks/common/data_components/FoodItemProperties.kt
package indi.muxin.food_talks.common.data_components

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import indi.muxin.food_talks.toRegistryName
import indi.muxin.food_talks.toResourceLocation
import net.minecraft.core.component.DataComponentType
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.ItemStack
import net.neoforged.neoforge.registries.RegisterEvent
import java.util.*
import java.util.function.Function


data class FoodItemProperties(
    val canAlwaysEat: Boolean,
    val convertsTo: Optional<ItemStack>,
    val chewSeconds: Float
)


private val codec: Codec<FoodItemProperties> = RecordCodecBuilder.create(
    Function {
        it.group(
            Codec.BOOL
                .fieldOf("can_always_eat")
                .forGetter(FoodItemProperties::canAlwaysEat),
            ItemStack.SINGLE_ITEM_CODEC
                .optionalFieldOf("using_converts_to")
                .forGetter(FoodItemProperties::convertsTo),
            Codec.FLOAT
                .fieldOf("eat_seconds")
                .forGetter(FoodItemProperties::chewSeconds)
        ).apply(it) { canAlwaysEat: Boolean, convertsTo: Optional<ItemStack>, chewSeconds: Float ->
            FoodItemProperties(
                canAlwaysEat,
                convertsTo,
                chewSeconds
            )
        }
    }
)
private val streamCodec: StreamCodec<RegistryFriendlyByteBuf, FoodItemProperties> = StreamCodec
    .composite(
        ByteBufCodecs.BOOL,
        FoodItemProperties::canAlwaysEat,

        ItemStack.STREAM_CODEC.apply { ByteBufCodecs.optional(it) },
        FoodItemProperties::convertsTo,

        ByteBufCodecs.FLOAT,
        FoodItemProperties::chewSeconds

    ) { canAlwaysEat: Boolean, convertsTo: Optional<ItemStack>, chewSeconds: Float ->
        FoodItemProperties(
            canAlwaysEat,
            convertsTo,
            chewSeconds
        )
    }


private val location: ResourceLocation = FoodItemProperties::class.java.getSimpleName().toRegistryName().toResourceLocation()

val FoodItemPropertiesDCType: DataComponentType<FoodItemProperties> = DataComponentType
    .builder<FoodItemProperties>()
    .persistent(codec)
    .networkSynchronized(streamCodec)
    .build()

fun RegisterEvent.RegisterHelper<DataComponentType<*>>.registerFoodItemProperties() {
    register(location, FoodItemPropertiesDCType)
}